examples/mamba.py

```diff
@@ -0,0 +1,472 @@
+        y = Tensor.einsum("bdn,bn->bd", x, C[:, :, i])
+      else:
+        y = Tensor.einsum("bdn,bdn->bd", x, C[:, :, :, i])
+    if i == u.shape[2] - 1:
+      last_state = x
+    # if y.is_complex():
+    #     y = y.real * 2
+    ys.append(y)
+  y = Tensor.stack(ys, dim=2)  # (batch dim L)
+  out = y if D is None else y + u * rearrange(D, "d -> d 1")
+  if z is not None:
+    out = out * z.silu()
+  return out if not return_last_state else (out, last_state)
+
+
+class MambaMixer:
+  def __init__(
+    self,
+    dim,
+    d_state=16,
+    d_conv=4,
+    expand=2,
+    dt_rank="auto",
+    dt_min=0.001,
+    dt_max=0.1,
+    dt_init="random",
+    dt_scale=1.0,
+    dt_init_floor=1e-4,
+    conv_bias=True,
+    bias=False,
+    layer_idx=None,
+  ):
+    self.dim = dim
+    self.d_state = d_state
+    self.d_conv = d_conv
+    self.expand = expand
+    self.d_inner = int(self.expand * self.dim)
+    self.dt_rank = math.ceil(self.dim / 16) if dt_rank == "auto" else dt_rank
+    self.layer_idx = layer_idx
+
+    self.in_proj = nn.Linear(self.dim, self.d_inner * 2, bias=bias)
+
+    self.conv1d = nn.Conv1d(
+      in_channels=self.d_inner,
+      out_channels=self.d_inner,
+      bias=conv_bias,
+      kernel_size=d_conv,
+      groups=self.d_inner,
+      padding=d_conv - 1,
+    )
+
+    self.x_proj = nn.Linear(self.d_inner, self.dt_rank + self.d_state * 2, bias=False)
+    self.dt_proj = nn.Linear(self.dt_rank, self.d_inner, bias=True)
+
+    # Initialize special dt projection to preserve variance at initialization
+    dt_init_std = self.dt_rank**-0.5 * dt_scale
+    if dt_init == "constant":
+      self.dt_proj.weight = Tensor.full(self.dt_proj.weight.shape, dt_init_std)
+    elif dt_init == "random":
+      self.dt_proj.weight = Tensor.uniform(
+        self.dt_proj.weight.shape, low=-dt_init_std, high=dt_init_std
+      )
+    else:
+      raise NotImplementedError
+
+    dt = (
+      (
+        Tensor.rand(self.d_inner) * (math.log(dt_max) - math.log(dt_min))
+        + math.log(dt_min)
+      )
+      .exp()
+      .maximum(dt_init_floor)
+    )
+    inv_dt = (
+      dt + (-((-dt).exp() - Tensor.ones(*dt.shape))).log()
+    )  # TODO: implement torch.expm1?
+
+    self.dt_proj.bias.assign(inv_dt)
+
+    # S4D real initialization
+    self.A_log = (
+      Tensor.arange(1, self.d_state + 1).repeat([self.d_inner, 1]).contiguous().log()
+    )
+
+    # D "skip" parameter
+    self.D = Tensor.ones(self.d_inner)  # Keep in fp32
+
+    self.out_proj = nn.Linear(self.d_inner, self.dim, bias=bias)
+
+  def __call__(self, hidden_states: Tensor, inference_params=None):
+    batch, seqlen, dim = hidden_states.shape
+
+    conv_state, ssm_state = None, None
+    if inference_params is not None:
+      conv_state, ssm_state = self._get_states_from_cache(inference_params, batch)
+      if inference_params.seqlen_offset > 0:
+        # The states are updated inplace
+        out, _, _ = self.step(hidden_states[:, -1:, :], conv_state, ssm_state)
```

Comment:
The `[:, -1:, :]` can be removed to match the original code. The removal was tested and it works fine